File: ios/Bonfire/Mobile/Users/FirebaseUserDatabase.swift
import Foundation
import RxSwift
import Firebase

class FirebaseUserDatabase: UserDatabase {
    let usersDB: FIRDatabaseReference

    init(usersDB: FIRDatabaseReference) {
        self.usersDB = usersDB
    }

    func observeUsers() -> Observable<Users> {
        return usersDB.rx_readValue().map(toUsers)
    }

    func observeUser(userID: String) -> Observable<User> {
        return usersDB.child(userID).rx_readValue().map(User.init)
    }

    func readUserFrom(userID: String) -> Observable<User> {
        return usersDB.child(userID).rx_readOnce().map(User.init)
    }

    func writeCurrentUser(user: User) {
        usersDB.child(user.identifier).setValue(user.asFirebaseValue())
    }

    private func toUsers(snapshot: FIRDataSnapshot) throws -> Users {
        let users = try snapshot.children.allObjects.map(asFIRDataSnapshot).map(User.init)
        return Users(users: users)
    }
}

File: android/app/src/main/java/com/novoda/bonfire/user/database/FirebaseUserDatabase.java
package com.novoda.bonfire.user.database;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.novoda.bonfire.rx.FirebaseObservableListeners;
import com.novoda.bonfire.user.data.model.User;
import com.novoda.bonfire.user.data.model.Users;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public class FirebaseUserDatabase implements UserDatabase {

    private final DatabaseReference usersDB;
    private final FirebaseObservableListeners firebaseObservableListeners;

    public FirebaseUserDatabase(FirebaseDatabase firebaseDatabase, FirebaseObservableListeners firebaseObservableListeners) {
        usersDB = firebaseDatabase.getReference("users");
        this.firebaseObservableListeners = firebaseObservableListeners;
    }

    @Override
    public Observable<Users> observeUsers() {
        return firebaseObservableListeners.listenToValueEvents(usersDB, toUsers());
    }

    @Override
    public Observable<User> readUserFrom(String userId) {
        return firebaseObservableListeners.listenToSingleValueEvents(usersDB.child(userId), as(User.class));
    }

    @Override
    public Observable<User> observeUser(String userId) {
        return firebaseObservableListeners.listenToValueEvents(usersDB.child(userId), as(User.class));
    }

    @Override
    public void writeCurrentUser(User user) {
        usersDB.child(user.getId()).setValue(user); //TODO handle errors
    }

    private Function<DataSnapshot, Users> toUsers() {
        return new Function<DataSnapshot, Users>() {
            @Override
            public Users apply(@NonNull DataSnapshot dataSnapshot) throws Exception {
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();
                List<User> users = new ArrayList<>();
                for (DataSnapshot child : children) {
                    User message = child.getValue(User.class);
                    users.add(message);
                }
                return new Users(users);
            }
        };
    }

    private <T> Function<DataSnapshot, T> as(final Class<T> tClass) {
        return new Function<DataSnapshot, T>() {
            @Override
            public T apply(@NonNull DataSnapshot dataSnapshot) throws Exception {
                return dataSnapshot.getValue(tClass);
            }
        };
    }
}
